From 55d00dbc3d2554887a2852fa2da724136788d3e9 Mon Sep 17 00:00:00 2001
From: Xiaolin Tang <tangxiaolin@loongson.cn>
Date: Wed, 14 Dec 2022 14:45:57 +0800
Subject: [PATCH 08/44] LoongArch: Hard Float Support for functions scalbn{f/
 }.

Use hardware Floating-point instruction fscaleb.{s/d}, fclass.{s/d} to
implement functions scalbn{f/ }.

Change-Id: I005dc6be86b2b1ebe87fc576b52442097370c4e0
---
 sysdeps/loongarch/fpu/s_scalbn.c  | 29 +++++++++++++++++++++++++++++
 sysdeps/loongarch/fpu/s_scalbnf.c | 29 +++++++++++++++++++++++++++++
 2 files changed, 58 insertions(+)
 create mode 100644 sysdeps/loongarch/fpu/s_scalbn.c
 create mode 100644 sysdeps/loongarch/fpu/s_scalbnf.c

diff --git a/sysdeps/loongarch/fpu/s_scalbn.c b/sysdeps/loongarch/fpu/s_scalbn.c
new file mode 100644
index 0000000000..c03e81a31e
--- /dev/null
+++ b/sysdeps/loongarch/fpu/s_scalbn.c
@@ -0,0 +1,29 @@
+/* scalbn().  LoongArch version.
+   Copyright (C) 2022 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <https://www.gnu.org/licenses/>.  */
+
+#define NO_MATH_REDIRECT
+#include <math.h>
+
+double
+__scalbn (double x, int fn)
+{
+  double tmp;
+  asm volatile ("movgr2fr.d \t%0, %1" : "=f" (tmp) : "r" (fn));
+  asm volatile ("fscaleb.d \t%0, %1, %2" : "=f" (x) : "f" (x), "f" (tmp));
+  return x;
+}
diff --git a/sysdeps/loongarch/fpu/s_scalbnf.c b/sysdeps/loongarch/fpu/s_scalbnf.c
new file mode 100644
index 0000000000..15e6428044
--- /dev/null
+++ b/sysdeps/loongarch/fpu/s_scalbnf.c
@@ -0,0 +1,29 @@
+/* scalbnf().  LoongArch version.
+   Copyright (C) 2022 Free Software Foundation, Inc.
+   This file is part of the GNU C Library.
+
+   The GNU C Library is free software; you can redistribute it and/or
+   modify it under the terms of the GNU Lesser General Public
+   License as published by the Free Software Foundation; either
+   version 2.1 of the License, or (at your option) any later version.
+
+   The GNU C Library is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+   Lesser General Public License for more details.
+
+   You should have received a copy of the GNU Lesser General Public
+   License along with the GNU C Library; if not, see
+   <https://www.gnu.org/licenses/>.  */
+
+#define NO_MATH_REDIRECT
+#include <math.h>
+
+float
+__scalbnf (float x, int fn)
+{
+  float tmp;
+  asm volatile ("movgr2fr.w \t%0, %1" : "=f" (tmp) : "r" (fn));
+  asm volatile ("fscaleb.s \t%0, %1, %2" : "=f" (x) : "f" (x), "f" (tmp));
+  return x;
+}
-- 
2.20.1

